feat(components): support specifying level padding in submenu

CarterLi opened the pull request:
Please make sure these boxes are checked before submitting your PR, thank you!

- [x] Make sure you follow contributing guide [English](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.en-US.md) | ([中文](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.zh-CN.md) | [Español](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.es.md) | [Français](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.fr-FR.md)).
- [x] Make sure you are merging your commits to `dev` branch.
- [x] Add some descriptions and refer to relative issues for your PR.

目前这个padding是hard code的20px，不能修改，而且是内联样式，也不方便使用全局样式覆盖。

Giwayume commented:
Maybe it would be better to have this controlled as a CSS variable?

```css
:root {
	--el-menu-item-level-padding: 20px;
	--el-menu-item-level: 1;
}
.el-menu-item {
	padding-left: calc(var(--el-menu-item-level-padding) * var(--el-menu-item-level));
}
```

Then the JS code changes to set `--el-menu-item-level` as an inline style instead of a specific padding value.

For reference, [there is a future spec called the inherit() function](https://github.com/w3c/csswg-drafts/issues/1026) which would allow this solution to work purely in CSS without JS. Won't be able to use this for a while, though.

```css
.el-sub-menu > .el-menu > .el-sub-menu,
.el-sub-menu > .el-menu > .el-menu-item-group {
	--el-menu-item-level: calc(1 + inherit(--el-menu-item-level));
}
```

CarterLi committed:
feat(components): menu add more controllable styles

1. add some css variables
2. support specifying level padding
3. small code refactor
4. deprecate `background-color` etc

CarterLi commented:
Done.

I also added more CSS variables to make el-menu more style-able. Please review! @Giwayume

Giwayume commented:
There are a couple of breaking changes if you compare the docs,
https://preview-7549-element-plus.surge.sh/en-US/component/menu.html#top-bar
https://element-plus.org/en-US/component/menu.html#collapse

Menu opening to the right instead of bottom, menu arrow direction incorrect:

![image](https://user-images.githubusercontent.com/4075314/167420503-8fb81c56-22e7-461a-8752-509b4cdea691.png)

Extra space to the left, and a horizontal line; too much padding:

![image](https://user-images.githubusercontent.com/4075314/167419835-489e0663-fd25-4fee-80e1-8bb14daa6529.png)

CarterLi commented:
> There are a couple of breaking changes if you compare the docs, https://preview-7549-element-plus.surge.sh/en-US/component/menu.html#top-bar https://element-plus.org/en-US/component/menu.html#collapse
> 
> Menu opening to the right instead of bottom, menu arrow direction incorrect:
> 
> ![image](https://user-images.githubusercontent.com/4075314/167420503-8fb81c56-22e7-461a-8752-509b4cdea691.png)
> 
> Extra space to the left, and a horizontal line; too much left padding, too little vertical padding:
> 
> ![image](https://user-images.githubusercontent.com/4075314/167419835-489e0663-fd25-4fee-80e1-8bb14daa6529.png)

As all css variables introduced in this PR have defaults as the old values, it should not break anything.

It is a bug. I will have a look tomorrow

CarterLi commented:
Fixed @Giwayume @YunYouJun

YunYouJun commented:
It works well. If you think it's a commonly used feature, should we provide a usage example?

CarterLi commented:
> It works well. If you think it's a commonly used feature, should we provide a usage example?

其他组件或多或少都提供了一些css变量用于定制化样式，但都没有提供示例什么的。CSS变量只在这里有提及：https://element-plus.org/zh-CN/guide/theming.html#%E9%80%9A%E8%BF%87-css-%E5%8F%98%E9%87%8F%E8%AE%BE%E7%BD%AE

我觉得也没必要。如果其他用户也有定制化样式的需求，他们自然会在调试器里找一找哪些样式提供了CSS变量。实在不行就 !important，活总不能不干

Giwayume commented:
It might be helpful to start documenting CSS variables on the individual components pages, sounds like a bigger task though.

Giwayume commented on packages/theme-chalk/src/common/var.scss:
```diff
@@ -996,9 +996,14 @@ $menu: map.merge(
     'item-font-size': getCssVar('font-size-base'),
     'item-hover-fill': getCssVar('color-primary-light-9'),
     'border-color': getCssVar('border-color'),
+    'top-padding': 20px,
```
Also, should this only affect padding-left? It has strange consequences on the arrow otherwise.

0px
![image](https://user-images.githubusercontent.com/4075314/167648114-5c1a1ebf-a149-49c9-a0c8-653b04bb803f.png)

40px (notice the arrow is aligned differently from sub-menu)
![image](https://user-images.githubusercontent.com/4075314/167648270-a6d70aab-3cec-49f8-897b-f62a27ce230d.png)

Giwayume replied:
I'm seeing weird behavior with this, since every time an `el-menu` is used as a child, it resets all of these variables.

Level padding is ignored in this case:

```
.el-menu {
  --el-menu-top-padding: 0;
}
```

![image](https://user-images.githubusercontent.com/4075314/167654312-3a1cbcdd-efbf-4c0c-b217-18bfa0536ade.png)

Arrows are misaligned:

```
.el-menu:not(.el-menu .el-menu) {
  --el-menu-top-padding: 0;
}
```
![image](https://user-images.githubusercontent.com/4075314/167654693-679c7fa7-2f46-42e6-af97-f9590d285dc7.png)

Giwayume replied:
@YunYouJun Would it be better to declare this variable on :root instead of `.el-menu`, so it can be inherited by sub-menu instead of overwritten?

CarterLi replied:
> Level padding is ignored in this case:

`--el-menu-level-padding` defaults to `var(--el-menu-top-padding)`, hence the same value.

Giwayume replied:
> > Level padding is ignored in this case:
> 
> `--el-menu-level-padding` defaults to `var(--el-menu-top-padding)`, hence the same value.

I think it's just the naming of `el-menu-top-padding` that was throwing me off here, because it suggests only the level: 0 item is affected. But in reality it is affecting all the items. You're looking for a way to configure that level: 0 item differently from all of the sub-menu children. I'd reverse the variable dependency.

```
'level-padding': 20px,
'top-level-padding': getCssVar('level-padding'),
```

CarterLi replied:
I'll make them separate.

YunYouJun replied:
> @YunYouJun Would it be better to declare this variable on :root instead of `.el-menu`, so it can be inherited by sub-menu instead of overwritten?

Yes, I think we can just make the necessary variables global. `:root` can also be placed in `menu.scss`. Because it only serves menu.

CarterLi commented:
And to be honest, properties like background-color, text-color, active-text-color should not exist.

`default-active`, `default-openeds` should be `v-model:active` `v-model:openeds`, and can be controlled by users.

Giwayume commented:
> And to be honest, properties like `background-color`, `text-color` should not exist.

I believe these properties were added before the theme moved to using CSS variables. At the end of the day they just set the CSS variables anyways, doesn't really hurt to leave them for backwards compatibility.

CarterLi commented:
> > And to be honest, properties like `background-color`, `text-color` should not exist.
> 
> I believe these properties were added before the theme moved to using CSS variables. At the end of the day they just set the CSS variables anyways, doesn't really hurt to leave them for backwards compatibility.

At least they should be deprecated

YunYouJun commented:
> It might be helpful to start documenting CSS variables on the individual components pages, sounds like a bigger task though.

I plan to write a script to extract them and display them in the document. But I haven't figured out the node api for sass yet.

YunYouJun commented:
> > > And to be honest, properties like `background-color`, `text-color` should not exist.
> > 
> > 
> > I believe these properties were added before the theme moved to using CSS variables. At the end of the day they just set the CSS variables anyways, doesn't really hurt to leave them for backwards compatibility.
> 
> At least they should be deprecated

I support deprecating them, of course we have to add tooltips for them and keep it until the next big release.

CarterLi commented:
Done @Giwayume @YunYouJun 

All css variables were moved to `:root` since they have the same problem as `level-padding`

Giwayume reviewed (approved):
This is looking good to me 👍

YunYouJun commented on packages/theme-chalk/src/menu.scss:
```diff
@@ -45,9 +45,12 @@
-@include b(menu) {
+:root {
+  // They are defined on :root so they can be inherited by sub-menu instead of overwritten
   @include set-component-css-var('menu', $menu);
```
Do all the variables here need to be moved to `:root`? Is there a css variable that can still be placed in `el-menu`?

CarterLi replied:
1. All variables should be in one place to avoid from confusion.
2. All variables exist for styling, which shares the same problem.

YunYouJun commented:
Thanks for your contribution.
